### ISSUE TITLE
[BUG] Dashboard shows 404 Not Found after upgrading from v1.3.1 to v1.3.2-rc1

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#6360]
**Describe the bug**

The dashboard shows 404 Not Found after upgrading from v1.3.1 to v1.3.2-rc1.

<img width="1914" alt="截圖 2024-08-16 下午4 10 19" src="https://github.com/user-attachments/assets/6b659af5-82ce-4d9c-8435-bbd88c7303af">

**To Reproduce**
1. Create a v1.3.1 cluster.
2. Upgrade to v1.3.2-rc1.
3. Access VIP with browser.

**Expected behavior**

Dashboard should display well.

**Support bundle**

**Environment**
 - Harvester ISO version: v1.3.1 -> v1.3.2-rc1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

With @starbops's help, we found the issue is caused by the argument change of `rke2-ingress-nginx-controller` pods:
- `--watch-ingress-without-class=true` argument is removed in RKE2 v1.28.12.
- The ingress `cattle-system/rancher-expose` has no `spec.ingressClassName` and it's ignored by the `rke2-ingress-nginx-controller`:
    ```
    2024-08-16T04:35:15.259654966Z I0816 04:35:15.259320       6 store.go:436] "Ignoring ingress because of error while validating ingress class" ingress="cattle-system/rancher-expose" error="ingress does not contain a valid IngressClass"
    ```
- The workaround is to edit the ingress and add an ingress class:
    ```
    $ kubectl edit ingress rancher-expose -n cattle-system
    
    <...>
    spec:
      ingressClassName: nginx  <--- add this
      rules:
      - http:
          paths:
          - backend:
              service:ot
                name: rancher
                port:
                  number: 80
            path: /
            pathType: Prefix
    <...>
    ```
- New v1.3.2-rc1 clusters are not affected. Even our bootstrap code creates the ingress without an ingress class, the resulting ingress has the `ingressClassName` field. The reason is unknown.
- This should affect v1.4-head too.

**Comment by starbops** [comment]
From the `kubectl.kubernetes.io/last-applied-configuration` annotation in the Ingress object, we know that the `.spec.ingressClassName` field was added on the fly:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"networking.k8s.io/v1","kind":"Ingress","metadata":{"annotations":{},"name":"rancher-expose","namespace":"cattle-system"},"spec":{"rules":[{"http":{"paths":[{"backend":{"service":{"name":"rancher","port":{"number":80}}},"path":"/","pathType":"Prefix"}]}}]}}
  creationTimestamp: "2024-08-17T09:13:15Z"
  generation: 1
  name: rancher-expose
  namespace: cattle-system
  resourceVersion: "103985"
  uid: f91c19b9-4732-4929-9ba2-74acb09d6b15
spec:
  ingressClassName: nginx
  rules:
  - http:
      paths:
      - backend:
          service:
            name: rancher
            port:
              number: 80
        path: /
        pathType: Prefix
status:
  loadBalancer: {}
```

It's done by the `DefaultIngressClass` admission controller plugin. The source can be found [here](https://github.com/kubernetes/kubernetes/blob/66df84bc1188b9fe224a91e881a0105b943d2545/plugin/pkg/admission/network/defaultingressclass/admission.go#L113-L124). There's also [documentation explaining the behavior](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#defaultingressclass). Basically, during the **object creation**, it will assign an ingress class name to the Ingress object if the object does not specify one and there is a default IngressClass object in the cluster. But this piece of code was added since 2020. Now that question is why this becomes an issue for us in an upgrade of v1.3.1 to v1.3.2-rc1.

Turns out it's because the `nginx` IngressClass object recently was added a new annotation: `ingressclass.kubernetes.io/is-default-class=true`. This is the `nginx` IngressClass object in observed in Harvester v1.3.1 clusters:

```yaml
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  annotations:
    meta.helm.sh/release-name: rke2-ingress-nginx
    meta.helm.sh/release-namespace: kube-system
  creationTimestamp: "2024-08-17T09:12:00Z"
  generation: 1
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: rke2-ingress-nginx
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: rke2-ingress-nginx
    app.kubernetes.io/part-of: rke2-ingress-nginx
    app.kubernetes.io/version: 1.9.6
    helm.sh/chart: rke2-ingress-nginx-4.9.100
  name: nginx
  resourceVersion: "1003"
  uid: 50c438d6-b73f-49ae-b218-bc2b5b9b9db2
spec:
  controller: k8s.io/ingress-nginx
```

On the other hand, the `nginx` IngressClass object in Harvester v1.3.2-rc1 clusters looks like the following:

```yaml
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  annotations:
    ingressclass.kubernetes.io/is-default-class: "true"
    meta.helm.sh/release-name: rke2-ingress-nginx
    meta.helm.sh/release-namespace: kube-system
  creationTimestamp: "2024-08-17T09:12:00Z"
  generation: 1
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: rke2-ingress-nginx
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: rke2-ingress-nginx
    app.kubernetes.io/part-of: rke2-ingress-nginx
    app.kubernetes.io/version: 1.10.1
    helm.sh/chart: rke2-ingress-nginx-4.10.102
  name: nginx
  resourceVersion: "103371"
  uid: 50c438d6-b73f-49ae-b218-bc2b5b9b9db2
spec:
  controller: k8s.io/ingress-nginx
```

The `rke2-ingress-nginx` chart incorporated in Harvester v1.3.1 is of version `4.9.100`. And it's `4.10.102` for Harvester v1.3.2-rc1. The key difference is exactly the introduction of THE default ingress class annotation (and it is set in the `rke2-ingress-nginx` HelmChart manifest during [RKE2 bootstrapping](https://github.com/rancher/rke2/blob/27989db0c83739c9f560e0567a3bf76177947122/pkg/bootstrap/bootstrap.go#L321).) Also, removing the `--watch-ingress-without-class=true` argument for the `nginx-ingress-controller` binary is one of the breaking changes for us. Relevant PRs are:

- rancher/rke2-charts#463
- rancher/rke2-charts#484
- rancher/rke2-charts#487

So, why does the issue NOT affect the newly installed v1.3.2-rc1 cluster? It's because of the **ordering**. For upgrading clusters, the `rancher-expose` Ingress object is already there, it doesn't matter whether the new IngressClass has the annotation or not. The `DefaultIngressCalss` admission controller only takes effect for newly created Ingress objects. But for newly installed clusters, the RKE2 is bootstrapped first, and then the Harvester chart is installed. That means the `nginx` IngressClass object with the annotation is created first. When the `rancher-expose` Ingress object is being created, the admission controller notice that there's a default IngressClass exists in the cluster, it will add the ingress class name to the Ingress object before persisting. That's the reason why newly installed v1.3.2-rc1 clusters aren't impacted.

For the conclusion, I would suggest to patch the `rancher-expose` Ingress object for the `.spec.ingressClassName=nginx` in our upgrade path. The other way is to **recreate** the very same Ingress object so that the admission controller can process it. I feel the `--watch-ingress-without-class=true` argument is for backward compatibility, so I wouldn't recommend adding it back when we have a better way of addressing the issue.

**Comment by khushboo-rancher** [comment]
This should get covered by release upgrade testing.
cc: @irishgordo

**Comment by irishgordo** [comment]
@khushboo-rancher - if we upgrade to rc1 from v1.3.1 it won't but if we upgrade to rc2 (yet to be released) from v1.3.1 it should be covered by upgrade testing then :smile:

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.3.2-rc2` (24/08/22).

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ <strong>Single node</strong>: After upgrade to v1.3.2-rc2, can access Harvester dashboard through VIP address $~~$</summary>

  ![image](https://github.com/user-attachments/assets/5c4c2470-d8e7-4c54-b11c-ec8d9ad38dd5)
  
  
  https://github.com/user-attachments/assets/7d671e19-5c84-460e-82c3-b2040078c66f



 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ <strong>Multiple nodes</strong>: After upgrade to v1.3.2-rc2, can access Harvester dashboard through VIP address $~~$</summary>

  ![image](https://github.com/user-attachments/assets/b330985e-baf2-4eec-b133-df0e1dea1a38)
  
  
  https://github.com/user-attachments/assets/eaf8783b-8fd0-4d71-b5b8-2fed8207201d




 </details>

### Test Information
* Test Environment: Single and three nodes harvester on kvm machines
* Harvester version: `v1.3.2-rc2` (24/08/22)

### Verify Steps
<details><summary>Upgrade from v1.3.1 to v1.3.2-rc2</summary>

  1. Prepare a v1.3.1 single or multiple nodes Harvester cluster
  2. Upgrade to v1.3.2-rc2
  3. Check can correctly upgrade to v1.3.2-rc2
  4. Open Harvester dashboard through VIP address

</details>
